File: apps/api/src/routes/devices.ts
import { Router, type Request, type Response } from "express";
import { getAllDevices, getDevice, getDeviceActuators, reorderDevices } from "../lib/sqlite.js";
import { broadcastDevices } from "../services/websocket.js";

export function createDevicesRouter(): Router {
  const router = Router();

  // Reorder devices — placed before /:id to avoid param conflict
  router.put("/order", (req: Request, res: Response) => {
    try {
      const { order } = req.body as { order: string[] };
      if (!Array.isArray(order)) {
        res.status(400).json({ ok: false, error: "order must be an array of device IDs" });
        return;
      }
      const orders = order.map((id, index) => ({ id, order: index }));
      reorderDevices(orders);
      broadcastDevices();
      res.json({ ok: true });
    } catch (err) {
      console.error("Error reordering devices", err);
      res.status(500).json({ ok: false, error: "Failed to reorder devices" });
    }
  });

  router.get("/", (_req: Request, res: Response) => {
    try {
      const devices = getAllDevices();
      res.json({ ok: true, devices });
    } catch (err) {
      console.error("Error fetching devices", err);
      res.status(500).json({ ok: false, error: "Failed to fetch devices" });
    }
  });

  router.get("/:id", (req: Request, res: Response) => {
    try {
      const device = getDevice(String(req.params.id));
      if (!device) {
        res.status(404).json({ ok: false, error: "Device not found" });
        return;
      }
      res.json({ ok: true, device });
    } catch (err) {
      console.error("Error fetching device", err);
      res.status(500).json({ ok: false, error: "Failed to fetch device" });
    }
  });

  router.get("/:id/actuators", (req: Request, res: Response) => {
    try {
      const device = getDevice(String(req.params.id));
      if (!device) {
        res.status(404).json({ ok: false, error: "Device not found" });
        return;
      }
      const actuators = getDeviceActuators(String(req.params.id));
      res.json({ ok: true, actuators });
    } catch (err) {
      console.error("Error fetching device actuators", err);
      res.status(500).json({ ok: false, error: "Failed to fetch device actuators" });
    }
  });

  return router;
}

File: apps/cortex/src/api/devices.py
"""
Device routes: GET /api/devices, GET /api/devices/:id, PUT /api/devices/order, actuators

Ported from apps/api/src/routes/devices.ts
"""

from fastapi import APIRouter
from pydantic import BaseModel


class ReorderRequest(BaseModel):
    order: list[str]


def create_devices_router(sqlite, ws_server) -> APIRouter:
    r = APIRouter()

    @r.put("/order")
    async def reorder_devices(body: ReorderRequest):
        try:
            orders = [{"id": id, "order": idx} for idx, id in enumerate(body.order)]
            sqlite.reorder_devices(orders)
            await ws_server.broadcast_devices(sqlite)
            return {"ok": True}
        except Exception as e:
            return {"ok": False, "error": f"Failed to reorder devices: {e}"}

    @r.get("")
    async def list_devices():
        try:
            devices = sqlite.get_all_devices()
            return {"ok": True, "devices": [d.to_dict() for d in devices]}
        except Exception as e:
            return {"ok": False, "error": f"Failed to fetch devices: {e}"}

    @r.get("/{device_id}")
    async def get_device(device_id: str):
        try:
            device = sqlite.get_device(device_id)
            if not device:
                return {"ok": False, "error": "Device not found"}
            return {"ok": True, "device": device.to_dict()}
        except Exception as e:
            return {"ok": False, "error": f"Failed to fetch device: {e}"}

    @r.get("/{device_id}/actuators")
    async def get_device_actuators(device_id: str):
        try:
            device = sqlite.get_device(device_id)
            if not device:
                return {"ok": False, "error": "Device not found"}
            actuators = sqlite.get_device_actuators(device_id)
            return {"ok": True, "actuators": [a.to_dict() for a in actuators]}
        except Exception as e:
            return {"ok": False, "error": f"Failed to fetch device actuators: {e}"}

    return r
